File: fundamentalsAlgo/JavaScript/valid_parenthese.js
// O(n) time and O(n) space
function validParenthesis(s) {
  const stack = [];
  const mapping = { ")": "(", "]": "[", "}": "{" };

  for (const c of s) {
    if (mapping[c]) {
      if (stack.length && stack[stack.length - 1] === mapping[c]) {
        stack.pop();
      } else {
        return false;
      }
    } else {
      stack.push(c);
    }
  }

  return stack.length === 0;
}

const out = validParenthesis("{{}}[]");
console.log(out);

module.exports = { validParenthesis };

File: fundamentalsAlgo/Python/valid_parenthese.py


# O(n) time and space

def valid_parenthesis(s):
    bracks=[]

    mapping={')': '(', ']':'[', '}':'{' }

    for c in s:
        if c in mapping:
            if bracks and bracks[-1]==mapping[c]:
                bracks.pop()
            else:
                return False
        else:
            bracks.append(c)
    
    if not bracks:
        return True
    else:
        return False
        

out=valid_parenthesis("{{}}[]")
print(out)